### PR TITLE
feat: enable plugin allowlist

### PR DESCRIPTION
**Pull request by wep21**
I enabled plugin blacklist for image transport publisher.
usage:
```
ros2 run usb_cam usb_cam_node_exe --ros-args -p image_raw.disable_pub_plugins:=["image_transport/compressed"]
```
```
❯ ros2 topic list
/camera_info
/image_raw
/image_raw/compressedDepth
/image_raw/theora
/parameter_events
/rosout
```

### CODE REVIEW

**Comment by wep21** [comment]
@gbiggs @jacobperron @ijnek Could you review this PR? Also, I wolud like to backport this change into humble.

File: image_transport/src/publisher.cpp [hunk 1]
```diff
@@ -110,9 +110,23 @@ Publisher::Publisher(
   impl_->base_topic_ = image_topic;
   impl_->loader_ = loader;
 
+  uint ns_len = node->get_effective_namespace().length();
```
Comment (ijnek) [inline image_transport/src/publisher.cpp]: Can ``param_base_name`` just be equal to ``base_topic``, rather than stripping off the namespace here?

**Comment by ijnek** [comment]
As this seems like a revival of a feature that existed in ROS 1, I'd like the maintainers to decide on these changes. I do agree that a publisher getting created for every image transport plugin installed is annoying, but a whitelist sounds like a better idea than a blacklist.

**Comment by wep21** [comment]
friendly ping @gbiggs @jacobperron

**Comment by gbiggs** [comment]
Please change this to use a whitelist rather than a blacklist, with an additional option to enable all publishers.

**Comment by wep21** [comment]
@gbiggs I changed this to whitelist at https://github.com/ros-perception/image_common/pull/264/commits/a6356ca4e6bd1bba3d2985ba505b7c38024374ff.

File: image_transport/src/publisher.cpp (hunk 1 above)
Comment (wep21) [reply image_transport/src/publisher.cpp]: @gbiggs @ijnek This comes from `image_transport_plugins`.
Which is better?

- use base topic name
  This looks like below.
 ```
/usb_cam:
  ros__parameters:
    image_raw:
      enable_pub_plugins:
      - image_transport/compressed
      format: jpeg
      jpeg_quality: 95
      png_level: 3
      tiff:
        res_unit: inch
        xdpi: -1
        ydpi: -1
```
- use whole topic name
  This looks like below.
```
/usb_cam:
  ros__parameters:
    image_raw:
      format: jpeg
      jpeg_quality: 95
      png_level: 3
      tiff:
        res_unit: inch
        xdpi: -1
        ydpi: -1
    /aaa/bbb/image_raw:
       enable_pub_plugins:
       - image_transport/compressed
```

Comment (ijnek) [reply image_transport/src/publisher.cpp]: Just the base topic name seems better.

**Comment by ijnek** [comment]
Should probably add all available transports to the whitelist is there are no plugins listed under ``enable_pub_plugins``. Otherwise, this change would break the default behavior of making all installed transports available (which many users depend on).

**Comment by wep21** [comment]
@ijnek okay, I will set all the plugins as a default parameter.

**Comment by ijnek** [comment]
@wep21 Apologies for the delay. I'm just reviewing again, and I think there is no need for a separate std::set and a std::vector, if you simplify:

```cpp
  for (const auto & lookup_name : loader->getDeclaredClasses()) {
    const std::string transport_name = erase_last_copy(lookup_name, "_pub");
    if (whitelist.count(transport_name)) {
      try {
        auto pub = loader->createUniqueInstance(lookup_name);
        pub->advertise(node, image_topic, custom_qos, options);
        impl_->publishers_.push_back(std::move(pub));
      } catch (const std::runtime_error & e) {
        RCLCPP_ERROR(
          impl_->logger_, "Failed to load plugin %s, error string: %s\n",
          lookup_name.c_str(), e.what());
      }
    }
  }
```

to:

```cpp
for (transport_name : whitelist_vec) {
  try {
    auto pub = loader->createUniqueInstance(transport_name);
    pub->advertise(node, image_topic, custom_qos, options);
    impl_->publishers_.push_back(std::move(pub));
  } catch (const std::runtime_error & e) {
    RCLCPP_ERROR(
      impl_->logger_, "Failed to load plugin %s, error string: %s\n",
      transport_name.c_str(), e.what());
  }
}
```

Please check that these changes do work, as I haven't tested it.

**Comment by wep21** [comment]
@ijnek I addressed your review [here](https://github.com/ros-perception/image_common/pull/264/commits/5a924947a24608a104aeaac8e6b6a07ef051b885). Could you review my changes? cc @gbiggs

**Review by ijnek — APPROVED** [review]

File: image_transport/src/publisher.cpp [hunk 2]
```diff
@@ -110,19 +110,36 @@ Publisher::Publisher(
...
+  } catch (const rclcpp::exceptions::ParameterAlreadyDeclaredException &) {
+    RCLCPP_DEBUG_STREAM(
+      node->get_logger(), param_base_name << ".enable_pub_plugins" << " was previously declared"
+    );
+    whitelist_vec =
+      node->get_parameter(
+      param_base_name +
+      ".enable_pub_plugins").get_value<std::vector<std::string>>();
+  }
+  for (size_t i = 0; i < whitelist_vec.size(); ++i) {
```
Comment (gbiggs) [inline image_transport/src/publisher.cpp]: What is the purpose of copying the contents of the vector into a set if you're then just going to iterate over it?

File: image_transport/src/publisher.cpp [hunk 3]
```diff
@@ -110,19 +110,36 @@ Publisher::Publisher(
...
-  for (size_t i = 0; i < blacklist_vec.size(); ++i) {
-    blacklist.insert(blacklist_vec[i]);
+  uint ns_len = node->get_effective_namespace().length();
+  std::string param_base_name = image_topic.substr(ns_len);
+  std::replace(param_base_name.begin(), param_base_name.end(), '/', '.');
+  if (param_base_name.front() == '.') {
+    param_base_name = param_base_name.substr(1);
   }
-
+  std::vector<std::string> whitelist_vec;
```
Comment (gbiggs) [inline image_transport/src/publisher.cpp]: Please use the term "allowlist" instead of "whitelist".

Comment (wep21) [reply image_transport/src/publisher.cpp]: @gbiggs I addressed [here](https://github.com/ros-perception/image_common/pull/264/commits/f24fa62231bb5623a6639560a8fd388de130e580)

File: image_transport/src/publisher.cpp (hunk 2 above)
Comment (wep21) [reply image_transport/src/publisher.cpp]: @gbiggs This is from original code, but I guess this removes duplicated plugins declared in the parameters.

**Review by ahcorde — APPROVED** [review]
@gbiggs can we merge this ?

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19161/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13730/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19910/)

**Review by ahcorde — CHANGES_REQUESTED** [review]
can you take a look to this windows failures ? https://ci.ros2.org/job/ci_windows/19910/console

File: image_transport/src/publisher.cpp [hunk 4]
```diff
@@ -110,19 +110,36 @@ Publisher::Publisher(
   impl_->base_topic_ = image_topic;
   impl_->loader_ = loader;
 
-  std::vector<std::string> blacklist_vec;
-  std::set<std::string> blacklist;
-  // nh.getParam(impl_->base_topic_ + "/disable_pub_plugins", blacklist_vec);
-  for (size_t i = 0; i < blacklist_vec.size(); ++i) {
-    blacklist.insert(blacklist_vec[i]);
+  uint ns_len = node->get_effective_namespace().length();
```
Comment (ahcorde) [inline image_transport/src/publisher.cpp]:
```suggestion
  auto ns_len = node->get_effective_namespace().length();
```

**Comment by wep21** [comment]
@ahcorde Thank you for suggesting the fix. Could you run ci again?

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19165/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13734/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19916/)

**Review by ahcorde — APPROVED** [review]

**Comment by Sushant-Chavan** [comment]
> @gbiggs @jacobperron @ijnek Could you review this PR? Also, I wolud like to backport this change into humble.

@wep21 Thank you for the PR. Do you also plan to merge these changes into `humble`?

**Comment by chrbrcknr** [comment]
Yes, it would be great to have this in `humble` too!

**Comment by juanluishortelano** [comment]
Yes, would also love to have this on humble

**Comment by noyzzz** [comment]
If you just want for example the compressed type you can enable it like this:

    start_gazebo_ros_image_bridge_back_cmd = Node(
        package='ros_gz_image',
        executable='image_bridge',
        name='ros_gz_image_bridge_back',
        arguments=['/camera_back/image_raw'],
        parameters=[{'camera_back.image_raw.enable_pub_plugins': ['image_transport/compressed']}],
        output='screen',
    )
note: for disabling everything except the raw one, you can't give an empty list, you just have to uninstall the image transport packages

**Pull request merged**